File: 1235-maximum-profit-in-job-scheduling/1235-maximum-profit-in-job-scheduling.cpp
class Solution {
public:
    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
        int n = startTime.size();
        
        vector<vector<int>> jobs;
        for(int i = 0; i < n; i++) {
            jobs.push_back({endTime[i], startTime[i], profit[i]});
        }
        
        sort(jobs.begin(), jobs.end());
        
        vector<int> dp(n + 1, 0);
        
        for(int i = 0; i < n; i++) {
            int end = jobs[i][0];
            int start = jobs[i][1];
            int p = jobs[i][2];
            
            int skip = dp[i];
            
            int low = 0, high = i - 1, j = -1;
            while(low <= high) {
                int mid = low + (high - low) / 2;
                if(jobs[mid][0] <= start) {
                    j = mid;
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
            
            int take = p;
            if(j != -1) {
                take += dp[j + 1];
            }
            
            dp[i + 1] = max(skip, take);
        }
        
        return dp[n];
    }
};

File: 1235-maximum-profit-in-job-scheduling/1235-maximum-profit-in-job-scheduling.py
class Solution:
    def jobScheduling(self, startTime, endTime, profit):
        jobs = []
        for i in range(len(startTime)):
            jobs.append((endTime[i], startTime[i], profit[i]))

        jobs.sort()

        n = len(jobs)
        dp = [0] * (n + 1)

        for i in range(n):
            end, start, p = jobs[i]

            skip = dp[i]

            low = 0
            high = i - 1
            j = -1
            while low <= high:
                mid = (low + high) // 2
                if jobs[mid][0] <= start:
                    j = mid
                    low = mid + 1
                else:
                    high = mid - 1

            if j >= 0:
                take = p + dp[j + 1]
            else:
                take = p

            dp[i + 1] = max(skip, take)

        return dp[n]